Mathlib/NumberTheory/ModularForms/EisensteinSeries/Basic.lean

```diff
@@ -0,0 +1,153 @@
+    one_div, moebiusEquiv, Matrix.SpecialLinearGroup.transpose, EquivLike.coe_coe,
+    Matrix.SpecialLinearGroup.toLin'_apply, Matrix.toLin'_apply', Matrix.mulVecLin_transpose,
+    Matrix.vecMulLinear_apply, Matrix.vecMul, Matrix.vec2_dotProduct, Int.cast_add, Int.cast_mul]
+  apply moebius_aux_lem k (A 0 0) (A 0 1) (A 1 0) (A 1 1) (i 0) (i 1) z ?_
+  apply UpperHalfPlane.denom_ne_zero A
+
+/--The permutation of a gammaSet given by multiplying by an element `Γ(N)` -/
+def gammaMoebiusFun (N : ℕ) (a b : ℤ) (γ : Gamma N) (f : gammaSet N a b) : (gammaSet N a b) := by
+  use Matrix.vecMul f.1 γ.1.1
+  simp only [gammaSet_mem', Matrix.vecMul, Matrix.vec2_dotProduct, Int.cast_add, Int.cast_mul]
+  have hγ := (Gamma_mem N _).1 γ.2
+  rw [hγ.1, hγ.2.2.1, hγ.2.2.2, hγ.2.1, f.2.1, f.2.2.1]
+  simp only [mul_one, mul_zero, add_zero, zero_add, true_and]
+  have := Matrix.SpecialLinearGroup.SL2_gcd (f.1 0) (f.1 1) f.2.2.2 γ
+  convert this
+  repeat{
+  simp_rw [Matrix.vecMul, Matrix.dotProduct, Matrix.vecCons]
+  simp only [Fin.sum_univ_two, Fin.cons_zero, Fin.cons_one]}
+
+lemma gammaMoebiusFun_eq_Moebequiv (N : ℕ) (a b : ℤ) (γ : Gamma N) (f : gammaSet N a b) :
+    (gammaMoebiusFun N a b γ f).1 = (moebiusEquiv γ.1 f.1) := by
+  simp only [gammaMoebiusFun, moebiusEquiv, Matrix.SpecialLinearGroup.transpose, EquivLike.coe_coe,
+    Matrix.SpecialLinearGroup.toLin'_apply, Matrix.toLin'_apply', Matrix.mulVecLin_transpose,
+    Matrix.vecMulLinear_apply]
+
+lemma gamma_left_inv (N : ℕ) (a b : ℤ ) (γ : Gamma N) (v : gammaSet N a b) :
+    gammaMoebiusFun N a b γ⁻¹ (gammaMoebiusFun N a b γ v) = v := by
+  simp_rw [gammaMoebiusFun, Matrix.vecMul_vecMul]
+  apply Subtype.ext
+  simp only [InvMemClass.coe_inv, SpecialLinearGroup.coe_inv, mul_adjugate,
+    SpecialLinearGroup.det_coe, one_smul, vecMul_one]
+
+
+lemma gamma_right_inv (N : ℕ) (a b : ℤ ) (γ : Gamma N) (v : gammaSet N a b) :
+    gammaMoebiusFun N a b γ (gammaMoebiusFun N a b γ⁻¹ v) = v := by
+  simp only [gammaMoebiusFun, InvMemClass.coe_inv, Matrix.SpecialLinearGroup.coe_inv,
+    Matrix.vecMul_vecMul]
+  apply Subtype.ext
+  simp only [adjugate_mul, SpecialLinearGroup.det_coe, one_smul, vecMul_one]
+
+/--The equivalence of gammaSets given by an element of `Γ(N)`-/
+def gammaMoebiusEquiv (N : ℕ) (a b : ℤ ) (γ : Gamma N) : (gammaSet N a b) ≃ (gammaSet N a b)
+    where
+  toFun := gammaMoebiusFun N a b γ
+  invFun := gammaMoebiusFun N a b γ⁻¹
+  left_inv v:= by
+    apply gamma_left_inv
+  right_inv v:= by
+    apply gamma_right_inv
+
+/-- The Eisenstein series of weight `k : ℤ` and level `Γ(N)`, for `N : ℕ`. -/
+def EisensteinLevelNtsum (k : ℤ) (N : ℕ) (a b : ℤ) : ℍ → ℂ := fun z => ∑' x : (gammaSet N a b),
+  (eise k z x)
+
+/-- The SlashInvariantForm defined by an Eisenstein series of weight `k : ℤ`, level `Γ(N)`,
+  and congruence condition given by `a b : ℤ`. -/
+def eisensteinLevelNSIF (N : ℕ) (k a b : ℤ) : SlashInvariantForm (Gamma N) k
```

Comment:
Is it difficult to prove that it is different from zero?

Reply:
So I don't know of an easy proof if this. I have a proof that for level 1 that some of them are not zero (i.e. E4 is not zero). But this uses the q-expansion which in turn uses a bunch of stuff that is going to be PRed later.  Once that is there I think I can probably give the q-expansion for all of these and hopefully one can show more of them are non-zero.

Reply:
In principle one can show non-vanishing by computing the constant term at a cusp, which is a bit easier than computing the whole q-expansion, but still very definitely non-trivial.